src/components/views/rooms/UserIdentityWarning.tsx

```diff
@@ -0,0 +1,308 @@
+/*
+Copyright 2024 New Vector Ltd.
+
+SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
+Please see LICENSE files in the repository root for full details.
+*/
+
+import React from "react";
+import {
+    CryptoEvent,
+    EventType,
+    KnownMembership,
+    MatrixEvent,
+    Room,
+    RoomStateEvent,
+    RoomMember,
+} from "matrix-js-sdk/src/matrix";
+import { logger } from "matrix-js-sdk/src/logger";
+
+import type { CryptoApi, UserVerificationStatus } from "matrix-js-sdk/src/crypto-api";
+import { _t } from "../../../languageHandler";
+import AccessibleButton, { ButtonEvent } from "../elements/AccessibleButton";
+import MemberAvatar from "../avatars/MemberAvatar";
+import { MatrixClientPeg } from "../../../MatrixClientPeg";
+import { SDKContext } from "../../../contexts/SDKContext";
+
+interface IProps {
+    // The current room being viewed.
+    room: Room;
+}
+
+interface IState {
+    // The current room member that we are prompting the user to approve.
+    currentPrompt: RoomMember | undefined;
+}
+
+// Does the given user's identity need to be approved?
+async function userNeedsApproval(crypto: CryptoApi, userId: string): Promise<boolean> {
+    const verificationStatus = await crypto.getUserVerificationStatus(userId);
+    return verificationStatus.needsUserApproval;
+}
+
+/**
+ * Displays a banner warning when there is an issue with a user's identity.
+ *
+ * Warns when an unverified user's identity has changed, and gives the user a
+ * button to acknowledge the change.
+ */
+export default class UserIdentityWarning extends React.Component<IProps, IState> {
```

Comment:
We favour the usage of hooks instead of react functional component (cf [code style](https://github.com/element-hq/element-web/blob/develop/code_style.md#react))

Reply:
I've refactored this.  I hope that I did it right.  I think the logic is harder to follow in this format, so any tips to make it clearer are welcome.